File: references/java-sdk/src/main/java/blue/language/sdk/internal/NodeObjectBuilder.java
package blue.language.sdk.internal;

import blue.language.model.Node;

import java.util.LinkedHashMap;
import java.util.Map;

public final class NodeObjectBuilder {

    private final Node node = new Node();

    private NodeObjectBuilder() {
    }

    public static NodeObjectBuilder create() {
        return new NodeObjectBuilder();
    }

    public NodeObjectBuilder type(String typeAlias) {
        node.type(typeAlias);
        return this;
    }

    public NodeObjectBuilder type(Class<?> typeClass) {
        node.type(TypeRef.of(typeClass).asTypeNode());
        return this;
    }

    public NodeObjectBuilder put(String key, Object value) {
        if (value instanceof Node) {
            node.properties(key, (Node) value);
            return this;
        }
        node.properties(key, new Node().value(value));
        return this;
    }

    public NodeObjectBuilder putNode(String key, Node value) {
        node.properties(key, value);
        return this;
    }

    public NodeObjectBuilder putStringMap(String key, Map<String, String> map) {
        Node dictionary = new Node().properties(new LinkedHashMap<String, Node>());
        if (map != null) {
            for (Map.Entry<String, String> entry : map.entrySet()) {
                String rawKey = entry.getKey();
                if (rawKey == null) {
                    continue;
                }
                String normalizedKey = rawKey.trim();
                if (normalizedKey.isEmpty()) {
                    continue;
                }
                String value = entry.getValue();
                dictionary.properties(normalizedKey, new Node().value(value));
            }
        }
        node.properties(key, dictionary);
        return this;
    }

    public NodeObjectBuilder putExpression(String key, String expression) {
        node.properties(key, new Node().value(expr(expression)));
        return this;
    }

    public Node build() {
        return node;
    }

    private static String expr(String expression) {
        if (expression == null) {
            return null;
        }
        String trimmed = expression.trim();
        if (trimmed.startsWith("${") && trimmed.endsWith("}")) {
            return trimmed;
        }
        return "${" + trimmed + "}";
    }
}

File: libs/sdk-dsl/src/lib/internal/node-object-builder.ts
import { BlueNode } from '@blue-labs/language';

import type { BlueValue, TypeInput } from '../types.js';
import { wrapExpression } from './expressions.js';
import { toBlueNode } from './node-input.js';
import { resolveTypeInput } from './type-input.js';

function ensureProperties(node: BlueNode): Record<string, BlueNode> {
  const existing = node.getProperties();
  if (existing) {
    return existing;
  }
  const created: Record<string, BlueNode> = {};
  node.setProperties(created);
  return created;
}

export class NodeObjectBuilder {
  private readonly node = new BlueNode().setProperties({});

  static create(): NodeObjectBuilder {
    return new NodeObjectBuilder();
  }

  type(typeInput: TypeInput): this {
    this.node.setType(resolveTypeInput(typeInput));
    return this;
  }

  put(key: string, value: BlueValue): this {
    ensureProperties(this.node)[key] = toBlueNode(value);
    return this;
  }

  putNode(key: string, value: BlueNode): this {
    ensureProperties(this.node)[key] = value.clone();
    return this;
  }

  putStringMap(
    key: string,
    map: Record<string, string> | Map<string, string>,
  ): this {
    const entries =
      map instanceof Map ? [...map.entries()] : Object.entries(map ?? {});
    const dictionary = new BlueNode().setProperties({});
    for (const [rawKey, rawValue] of entries) {
      const normalizedKey = rawKey.trim();
      if (normalizedKey.length === 0) {
        continue;
      }
      dictionary.addProperty(normalizedKey, new BlueNode().setValue(rawValue));
    }
    ensureProperties(this.node)[key] = dictionary;
    return this;
  }

  putExpression(key: string, expression: string): this {
    ensureProperties(this.node)[key] = new BlueNode().setValue(
      wrapExpression(expression),
    );
    return this;
  }

  build(): BlueNode {
    return this.node.clone();
  }
}
